apps/hash-frontend/src/pages/login.page.tsx

```diff
@@ -126,7 +114,8 @@ const LoginPage: NextPageWithLayout = () => {
             }
 
             updateActiveWorkspaceAccountId(authenticatedUser.accountId);
-            void router.push("/");
+
+            void router.push(returnTo ?? flow.return_to ?? "/");
```

Comment:
We should address this to avoid `https://hash.ai/login?return_to=https://evil.com`, where https://evil.com could be a copy of the password screen and a 'try again' message etc.

one option would be to construct the URL afresh, and just take the pathname. 

With pathnames alone, we also need to be careful of paths with double slashes, which are interpreted as 'a URL with whatever protocol the current page uses', e.g. `http://localhost:3000/login?return_to=//google.com` would redirect to `http://google.com`. NextJS has protection for this but our own error will provide more detail in Sentry, and guard against us swapping out the router / NextJS behavior changing. More [here](https://snyk.io/blog/url-confusion-vulnerabilities).

```ts
  const possiblyMaliciousRedirect =
    typeof router.query.return_to === "string"
      ? router.query.return_to
      : undefined;

  const redirectUrl = possiblyMaliciousRedirect
    ? new URL(possiblyMaliciousRedirect, frontendUrl)
    : undefined;
  const redirectPath = redirectUrl?.pathname;

  // this isn't strictly necessary since we're only going to take the pathname, but useful to have the error reported
  if (redirectUrl && redirectUrl.origin !== frontendUrl) {
    throw new Error(
      `Someone tried to pass an external URL as a redirect: ${possiblyMaliciousRedirect}`,
    );
  }

  // next/router will error if these are included in the URL, but this makes the error more useful
  if (
    redirectPath &&
    (redirectPath.includes("\\") || redirectPath.includes("//"))
  ) {
    throw new Error(
      `Someone tried to pass a malformed URL as a redirect: ${possiblyMaliciousRedirect}`,
    );
  }
```

Reply:
Thanks for the detailed suggestion Ciaran, have addressed in https://github.com/hashintel/hash/pull/3105/commits/9a52caa2969762eedbce3b7551cf27608d423fd5. I'm wondering if we really do want to error when this happens though, when we could just ignore the query param when it's malformed? I appreciate we would lose the error messages, but we could probably log it in Sentry in some other way if we want to track when this is has occurred?

Reply:
Arguments for just throwing:
1. We don't have to add explicit `Sentry.captureException` calls here
2. This shouldn't happen unless someone has been sent here via a malicious link, and it's arguably better for them to know that there's something wrong with the source of the link

Arguments against throwing:
1. Users who come via a malicious link might think there's something wrong with the app generally and be put off using it

I would lean towards throwing to make it obvious something has gone wrong but we could change it later. cc @nonparibus for any views

Reply:
I think the user should just be redirected to a safe page (e.g. their default location, post-login -- which might be something users can specify down the line, and default to 'home' for now). I don't think it's critical they're informed a redirect failed.

The failed redirect must, however, be logged our side as it is a security event.